### ISSUE TITLE
What's the difference between accuracy in SED's metrics and binary_accuracy in tensorflow?

### ISSUE THREAD

**Issue by Lamyaaa** [TUT-ARG/sed_eval#6]
Great library! 
I have use this library for a while, but still I can't tell the difference between segment-based (set segment length to 1 frame) [accurary](http://tut-arg.github.io/sed_eval/sound_event.html#) used in SED and [binary accuracy](https://github.com/tensorflow/tensorflow/blob/r1.3/tensorflow/contrib/keras/python/keras/metrics.py) in tensorflow.  My understanding is that 1 frame segment-based equal to binary accuracy in tensorflow, since they both calculate the acc in each frame. However, these two value are not equal in my model. I just don't know why.

The accuracy I mean is 
> accuracy = ( TP + TN ) / ( TP + TN + FP + FN )

Binary accuracy in tensorflow is 
> def binary_accuracy(y_true, y_pred):
> return K.mean(K.equal(y_true, K.round(y_pred)), axis=-1)

Two formula above actually mean the same thing. My output is frame by frame, and each frame has num_of_class scalars to present existence of each class.

Please tell me whether I misunderstand it. Thanks!

**Comment by toni-heittola** [comment]
You are right they should produce same result for same size data. However, because sed_eval works with event lists which are generated into binary matrices (size of the matrix is decided by the last offset in time line), you might end up having different size of matrices in sed_eval than you had in Tensorflow.

Usually this is not an issue, as accuracy is not recommended metric for SED as it is inflated by zero columns in the binary matrix. Usually used metrics, F-score and ER, are not affected by this. Originally accuracy was included in the sed_eval toolbox for the sake of completeness. 

This issue can be fixed quite straightforward manner by introducing extra parameter to the evaluate method and specifying the length of the evaluated time for each evaluated item. I will commit fix for this a bit later.

**Comment by Lamyaaa** [comment]
Thanks for your reply!

Agree with you accuracy is not a important metric for SED. But I'm still confused on the question.

Actually I have the same time length of all my samples (i.e., each T-frame sequence as a sentence). Though accuracy in sed_eval not equal to binary accuracy in tensorflow, they are close to each other (e.g. 0.951349 | 0.950152)  and have the almost same variation trend.

By the way, I found a sad phenomenon that there is a high mismatching between binary cross entropy and ER/F1 metrics in sed_eval. Usually, bce in validate set begin to rise while ER/F1 still getting good. Does this relate to ER/F1 are calculate over whole data set, instead of each sample?

**Comment by toni-heittola** [comment]
I added now a new parameter, evaluated_length_seconds, for evaluate method in SegmentBasedMetric (see commit e8b96903f52d127bb14bb5eafe0efabee8913bc5). With this parameter you can give the length of evaluated audio signal, and the binary matrices will be generated to this length before using them to calculate metrics. You can see if this will give you exactly same numbers.

Since you have such a small difference between sed_eval and Tensorflow metric values, either your system outputs activity mostly until the end of all input signal or your reference event lists have most of time events which have offset at the end of signal.

Regarding on mismatch between binary cross entropy and ER/F1: In ideal case when binary cross entropy is low also ER should be low, and at the same time F1 high. However, without knowing more about your setup (what type of data, what metric parameters) I cannot dig deeper on this matter. What I can say right away, is that ER should not be used at frame-level.
